src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -670,215 +768,169 @@ object OrderedRVD {
-    hintPartitioner: OrderedRVDPartitioner
-  ): OrderedRVD = coerce(typ, rdd, None, Some(hintPartitioner))
-
-  def coerce(
-    typ: OrderedRVDType,
-    rdd: RDD[RegionValue],
-    fastKeys: RDD[RegionValue],
-    hintPartitioner: OrderedRVDPartitioner
-  ): OrderedRVD = coerce(
-    typ,
-    rdd,
-    Some(fastKeys),
-    Some(hintPartitioner))
-
-  def coerce(
-    typ: OrderedRVDType,
-    rdd: RDD[RegionValue],
-    fastKeys: Option[RDD[RegionValue]],
-    hintPartitioner: Option[OrderedRVDPartitioner]
+    fastKeys: Option[RDD[RegionValue]]
   ): OrderedRVD = coerce(
     typ,
     ContextRDD.weaken[RVDContext](rdd),
-    fastKeys.map(ContextRDD.weaken[RVDContext](_)),
-    hintPartitioner)
+    fastKeys.map(ContextRDD.weaken[RVDContext](_))
+    )
 
   def coerce(
     typ: OrderedRVDType,
     crdd: ContextRDD[RVDContext, RegionValue]
-  ): OrderedRVD = coerce(typ, crdd, None, None)
+  ): OrderedRVD = coerce(typ, crdd, None)
 
   def coerce(
     typ: OrderedRVDType,
     crdd: ContextRDD[RVDContext, RegionValue],
     fastKeys: ContextRDD[RVDContext, RegionValue]
-  ): OrderedRVD = coerce(typ, crdd, Some(fastKeys), None)
+  ): OrderedRVD = coerce(typ, crdd, Some(fastKeys))
+
+  def coerce(
+    typ: OrderedRVDType,
+    crdd: ContextRDD[RVDContext, RegionValue],
+    fastKeys: Option[ContextRDD[RVDContext, RegionValue]]
+  ): OrderedRVD = {
+    val keys = fastKeys.getOrElse(getKeys(typ, crdd))
+    makeCoercer(typ, keys).coerce(typ, crdd)
+  }
 
   def makeCoercer(
     fullType: OrderedRVDType,
-    crdd: ContextRDD[RVDContext, RegionValue],
-    hintPartitioner: Option[OrderedRVDPartitioner]): RVDCoercer = {
-    val sc = crdd.sparkContext
+    // keys: RDD[RegionValue[fullType.kType]]
+    keys: ContextRDD[RVDContext, RegionValue]
+   ): RVDCoercer = {
+    type CRDD = ContextRDD[RVDContext, RegionValue]
+    val sc = keys.sparkContext
 
     val emptyCoercer: RVDCoercer = new RVDCoercer(fullType) {
-      def _coerce(typ: OrderedRVDType, crdd: ContextRDD[RVDContext, RegionValue]): OrderedRVD = empty(sc, typ)
+      def _coerce(typ: OrderedRVDType, crdd: CRDD): OrderedRVD = empty(sc, typ)
     }
 
-    if (crdd.partitions.isEmpty)
-      return emptyCoercer
-
-    // keys: RDD[RegionValue[kType]]
-    val keys = crdd
-
-    val pkis = getPartitionKeyInfo(fullType, keys)
+    val pkis = getKeyInfo(fullType, keys)
 
     if (pkis.isEmpty)
       return emptyCoercer
 
-    val partitionsSorted = (pkis, pkis.tail).zipped.forall { case (p, pnext) =>
-      val r = fullType.pkType.ordering.lteq(p.max, pnext.min)
-      if (!r)
-        log.info(s"not sorted: p = $p, pnext = $pnext")
-      r
+    val bounds = pkis.map(_.interval).toFastIndexedSeq
+    val pkBounds = bounds.map(_.coarsen(fullType.partitionKey.length))
+    def orderPartitions = { crdd: CRDD =>
+      val pids = pkis.map(_.partitionIndex)
+      if (pids.isSorted && crdd.getNumPartitions == pids.length)
+        crdd
+      else
+        crdd.reorderPartitions(pids)
```

Comment:
I bet we'll find it helpful to have a log message including `pids` and `crdd.getNumPartitions` when we take the alternate path here.

Reply:
Do we want to log something as potentially large as `pids`?

Reply:
I could just have two log messages, one if `crdd` has fewer partitions than `keys` did, and the other if the partitions in `keys` were out of order.

Reply:
good point, I think your suggestion is a better route.